### PR TITLE
extracting list from CustomerGateways to match create

### LINKED ISSUE #24747: ec2_customer_gateway returns different values depending if CGW exists already or not
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
ec2_customer_gateway

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

```

##### CONFIGURATION
standard

##### OS / ENVIRONMENT
Ubuntu, AWS

##### SUMMARY
Creating CGW vs reporting on existing CGW returns different results.

##### STEPS TO REPRODUCE
Sample task:
```yaml
- name: Create Spoke Customer Gateway
  ec2_customer_gateway:
    bgp_asn: "{{ asn_spoke }}"
    ip_address: "{{ item.public_ip }}"
    name: cgw-spoke
    region: "{{ aws_region }}"
    aws_access_key: "{{ ec2_access_key }}"
    aws_secret_key: "{{ ec2_secret_key }}"
  with_items: "{{ eip_spoke.results }}"
  register: cgw_spoke
```

##### EXPECTED RESULTS
On the first run, a dictionary "customer_gateway" is returned. On subsequent runs, list "customer_gateway" and list "customer_gateways" are returned. This creates issues when attempting to code for create and already created scenarios.

##### ACTUAL RESULTS

Running the above task the first time returns the following result:
```
"gateway": {
                    "customer_gateway": {
                        "bgp_asn": "65000", 
                        "customer_gateway_id": "cgw-1d05de03", 
                        "ip_address": "52.27.234.211", 
                        "state": "available", 
                        "type": "ipsec.1"
                    }
}
```

Running the above task with the same information will return the following result:
```
"gateway": {
                    "customer_gateway": [
                        {
                            "bgp_asn": "65000", 
                            "customer_gateway_id": "cgw-1d05de03", 
                            "ip_address": "52.27.234.211", 
                            "state": "available", 
                            "tags": [
                                {
                                    "key": "Name", 
                                    "value": "cgw-spoke"
                                }
                            ], 
                            "type": "ipsec.1"
                        }
                    ], 
                    "customer_gateways": [
                        {
                            "bgp_asn": "65000", 
                            "customer_gateway_id": "cgw-1d05de03", 
                            "ip_address": "52.27.234.211", 
                            "state": "available", 
                            "tags": [
                                {
                                    "key": "Name", 
                                    "value": "cgw-spoke"
                                }
                            ], 
                            "type": "ipsec.1"
                        }
                    ]
}

```

### PR DESCRIPTION
**Pull request by mcaulifn**
##### SUMMARY
Fixes #24747 
Uses list 0 from dict CustomerGateways in dict CustomerGateway to match the results returned during the create operation.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 3.5.2 (default, Nov 17 2016, 17:05:23) [GCC 5.4.0 20160609]

```

##### ADDITIONAL INFORMATION
Output of 
````
- debug: var=cgw_spoke.results[0].gateway
````
Before:
```
ok: [localhost] => {
    "cgw_spoke.results[0].gateway": {
        "customer_gateway": [
            {
                "bgp_asn": "65000", 
                "customer_gateway_id": "cgw-ce33e8d0", 
                "ip_address": "34.209.88.17", 
                "state": "available", 
                "tags": [
                    {
                        "key": "Name", 
                        "value": "cgw-spoke"
                    }
                ], 
                "type": "ipsec.1"
            }
        ], 
        "customer_gateways": [
            {
                "bgp_asn": "65000", 
                "customer_gateway_id": "cgw-ce33e8d0", 
                "ip_address": "34.209.88.17", 
                "state": "available", 
                "tags": [
                    {
                        "key": "Name", 
                        "value": "cgw-spoke"
                    }
                ], 
                "type": "ipsec.1"
            }
        ], 
        "response_metadata": {
            "http_headers": {
                "content-_type": "text/xml;charset=UTF-8", 
                "date": "Mon, 22 May 2017 16:34:26 GMT", 
                "server": "AmazonEC2", 
                "transfer-_encoding": "chunked", 
                "vary": "Accept-Encoding"
            }, 
            "http_status_code": 200, 
            "request_id": "fa980219-d39d-4cff-84be-25d0488c53c3", 
            "retry_attempts": 0
        }
    }, 
    "changed": false
}
```
After:
````
ok: [localhost] => {
    "cgw_spoke.results[0].gateway": {
        "customer_gateway": {
            "bgp_asn": "65000", 
            "customer_gateway_id": "cgw-ce33e8d0", 
            "ip_address": "34.209.88.17", 
            "state": "available", 
            "tags": [
                {
                    "key": "Name", 
                    "value": "cgw-spoke"
                }
            ], 
            "type": "ipsec.1"
        }, 
        "customer_gateways": [
            {
                "bgp_asn": "65000", 
                "customer_gateway_id": "cgw-ce33e8d0", 
                "ip_address": "34.209.88.17", 
                "state": "available", 
                "tags": [
                    {
                        "key": "Name", 
                        "value": "cgw-spoke"
                    }
                ], 
                "type": "ipsec.1"
            }
        ], 
        "response_metadata": {
            "http_headers": {
                "content-_type": "text/xml;charset=UTF-8", 
                "date": "Mon, 22 May 2017 16:06:38 GMT", 
                "server": "AmazonEC2", 
                "transfer-_encoding": "chunked", 
                "vary": "Accept-Encoding"
            }, 
            "http_status_code": 200, 
            "request_id": "4462d648-79cf-4b84-86f1-88e0d0512e5c", 
            "retry_attempts": 0
        }
    }, 
    "changed": false
}
````

### CODE REVIEW

**Comment by MichaelBaydoun** [comment]
Agree the output should be fixed to be consistent.  Can we change the returned results without introducing a potentially breaking change for some users?  Thoughts?

**Comment by mcaulifn** [comment]
Add in another variable or modify CustomerGateway or CustomerGateways? I'm not sure how other tasks have been written to support either situation.

**Comment by s-hertel** [comment]
Since this module is preview (as opposed to stableinterface) maybe changing the returned results is okay? But checking out the change running this task:
```
    - name: create customer gateway
      ec2_customer_gateway:
        bgp_asn: "{{ bgp_asn }}"
        ip_address: "{{ ip }}"
        name: "{{ name }}"
        region: us-east-1
      register: cgw
```
results in
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_C0c5Ym/ansible_module_ec2_customer_gateway.py", line 272, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_C0c5Ym/ansible_module_ec2_customer_gateway.py", line 229, in main
    existing['CustomerGateway'] = existing['CustomerGateways'][0]
IndexError: list index out of range

fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_C0c5Ym/ansible_module_ec2_customer_gateway.py\", line 272, in <module>\n    main()\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_C0c5Ym/ansible_module_ec2_customer_gateway.py\", line 229, in main\n    existing['CustomerGateway'] = existing['CustomerGateways'][0]\nIndexError: list index out of range\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
	to retry, use: --limit @/Users/shertel/Workspace/ansible/my_playbooks/ec2_customer_gateway/test.yml

PLAY RECAP *************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```
Which doesn't happen with the current ec2_customer_gateway.py.

**Comment by mcaulifn** [comment]
I found the issue and corrected. I was not properly handling an empty list. I found two ways of addressing:

- Change the test for existing CGWs to look at CustomerGateways and only copy the data to CustomerGateway if needed (commit a3f2090d)
- Try/Except logic that could copy the data if it existed or created an empty list if it did not (example below)

````
    try:
        existing['CustomerGateway'] = existing['CustomerGateways'][0]
    except:
        existing['CustomerGateway'] = existing['CustomerGateways']
````

I went with the first option as that seemed the more correct. Either should work.

File: lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py [hunk 1]
```diff
@@ -226,20 +226,21 @@ def main():
     existing = gw_mgr.describe_gateways(module.params['ip_address'])
     # describe_gateways returns a key of CustomerGateways where as create_gateway returns a
     # key of CustomerGateway. For consistency, change it here
-    existing['CustomerGateway'] = existing['CustomerGateways']
+    #existing['CustomerGateway'] = existing['CustomerGateways'][0]
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py]: You can remove this.

**Comment by s-hertel** [comment]
LGTM.
I'm unsure about the return contents in general for this module. For reference here's creating:
```
changed: bool,
gateway: dict containing (customer_gateway: dict, 
                          response_metadata: dict, ),
name: dict
```
And here is running a second time:
```
changed: bool,
gateway: dict containing (customer_gateway: dict, 
                          response_metadata: dict,
                          customer_gateways: list of one dict),

```
Since customer_gateways seems to never be more than one dict I'm not sure how it's helpful to have it at all. Also, name is the response from setting a tag but doesn't actually contain the tag's name and is only returned during creation. During creation there is no good way to get the tags (though it's in gateway during subsequent runs. It just seems a little confusing. @MichaelBaydoun Do you have a vision for how you want the returns to look?

**Comment by willthames** [comment]
@s-hertel, I noticed the missing tags too - the easiest way to address that would be to call `describe_customer_gateways` after calling `tag_cgw_name` on the newly created customer gateway.

All the fixes such as removing CustomerGateways from the dict after assigning the first element to CustomerGateway can be done after that describe_customer_gateways call.

**Comment by s-hertel** [comment]
shipit

**Review by s-hertel — APPROVED** [review]

**Review by s-hertel — CHANGES_REQUESTED** [review]
Hey, we can't merge this into stable-2.3. Sorry for not noticing that. Can you please rebase on devel and request to merge into 2.4?

**Comment by mcaulifn** [comment]
rebase wiped out some changes, will be re-committing shortly

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Comment by pilou-** [comment]
@willthames @MichaelBaydoun could you review the updates ? Depending on the result, `needs_revision` comment or a `shipit` comment could be added.

**Comment by MichaelBaydoun** [comment]
As @s-hertel pointed out, since not stable interface, I'm ok with not maintaining backwards compatibility, shipit.

**Pull request merged**